### ISSUE TITLE
`const fn`

### ISSUE THREAD

**Issue by Kixunil** [rust-bitcoin/rust-bitcoin#1174]
After MSRV bump we can start making functions `const`. I think we have many of them that can and should be `const` (like the newly added newtypes). I forgot about this when discussing MSRV bump but I don't mind since this is probably large anyway. We can't just mindlessly add them wherever compiler doesn't complain, we have to answer this question for each fn: "can we guarantee that this function will never call into OS or mutate global state?"

**Comment by apoelstra** [comment]
Agreed! Though I'm sure there are a lot of easy cases (e.g. most of the stuff in `Address` or `bip32` are simple pure functions)

**Comment by tcharding** [comment]
const functions cannot use an `if` statement either.

**Comment by Kixunil** [comment]
Depends on Rust version and there are funny workarounds like this:

```rust
let x = [val_if_true, val_if_false];
x[cond as usize]
```

**Comment by tcharding** [comment]
Interesting, cheers.

**Comment by tcharding** [comment]
> const functions cannot use an if statement either.

Well that is what the compiler told me the other day but today I can get this to compile
```rust
            /// Create an object from a given signed 64-bit integer
            #[inline]
            pub const fn from_i64(init: i64) -> Option<$name> {
                if init >= 0 {
                    $name::from_u64(init as u64)
                } else {
                    None
                }
            }
```
After making `from_u64` const as well.

**Comment by tcharding** [comment]
I edited your issue description @Kixunil, please remove the comment if you don't like me editing your text - no hard feelings.

**Comment by tcharding** [comment]
getter methods of wrapper types are a good candidate (e.g., methods that contain a single line `self.0`)

**Comment by Kixunil** [comment]
Constructors are better candidates, since one usually wants to write `STATIC FOO: Foo = Foo::new(42);`

**Comment by tcharding** [comment]
We got a hot tip in https://github.com/rust-bitcoin/rust-bitcoin/issues/2814#issue-2325229512

> May be, to prevent this in future, you may add `clippy::missing_const_for_fn` lint to configuration file?

**Comment by tcharding** [comment]
I don't see the advantage of const'ing functions that are not constructors, when does one need a function that is not a constructor to be const? It seems overly restrictive and seems to fly against all the `non_exhaustive` and error stuff we've done to make it possible to maintain the lib in a non-api-breaking way.

**Comment by apoelstra** [comment]
>when does one need a function that is not a constructor to be const

When you are defining constructors of your own :P.

**Comment by tcharding** [comment]
lol, obvious once I read it.

**Comment by apoelstra** [comment]
> seems to fly against all the `non_exhaustive` and error stuff we've done to make it possible to maintain the lib in a non-api-breaking way

For sure -- but there's always a tradeoff here. If "maintenance without breakage" were our only goal then we could create an opaque `RustBitcoinType`, make every single function receive and return a `RustBitcoinType`, and implement our own runtime typing system and library inside of there. Internally we could support vectors which would allow us to change the "number of parameters" that functions received, and support poisoned values in case somebody called a method that we later decided not to support, or whatever.

The tradeoff is basically "every promise we make to our users is a promise we can't later break" vs "every promise we *don't* make is one that users have to code around the possibility of being broken".

My view is that constness, while not having a *zero* cost, has a near-zero cost in 95% of cases. We just need to be a bit careful about identifying those cases. And from a user point of view the benefit is pretty big because it lets them construct increasingly elaborate constant values and compile-time assertions.

**Comment by Kixunil** [comment]
> My view is that constness, while not having a zero cost, has a near-zero cost in 95% of cases. We just need to be a bit careful about identifying those cases.

I agree with this. The only problem are things that might need generics later but that's also limited to generics with traits. And in ten years even those will be all const. Strictly speaking, every function in this library should be const long term, possibly except bitcoinconsensus FFI calls. (Also `Read`/`Write` should be `const` IFF their impls are.)

**Comment by tcharding** [comment]
Making a function `const` is a non-breaking change, right? But we should audit the leaf crates for candidate functions before 1.0-ing them, right?

Can someone add the 1.0 label if this is correct?
